File: island-count.java
// dfs soln
// Time Complexity : O(M * N)
// Space Complexity : O(M * N)
// Did this code successfully run on Leetcode : Yes

class Solution {
    int m; int n;
    int count = 0;
    int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};
    public int numIslands(char[][] grid) {
        if(grid == null || grid.length == 0)
            return 0;
        m = grid.length; n = grid[0].length;
        for(int i = 0; i < m; i++){
            for(int j = 0; j< n; j++){
                if(grid[i][j] == '1'){
                    count++;
                    dfs(grid, i , j);
                }
            }
        }
        return count;
    }
    private void dfs(char[][] grid, int row, int col){
        //base
        if(row < 0 || row == m || col < 0 || col == n || grid[row][col] != '1')
            return;
        //logic
        grid[row][col] = '0';
        for(int[] dir: dirs){
            int r = dir[0] + row;
            int c = dir[1] + col;
            dfs(grid, r, c);
        }
    }
}


/*
// bfs soln
// Time Complexity : O(M * N)
// Space Complexity : O(min(M, N))
// Did this code successfully run on Leetcode : Yes

class Solution {
    public int numIslands(char[][] grid) {
        if(grid == null || grid.length == 0)
            return 0;
        Queue<Integer> q = new LinkedList<>(); 
        int m = grid.length; int n = grid[0].length;
        int count = 0;
        for(int i = 0; i < m; i++){
            for(int j = 0; j< n; j++){
                if(grid[i][j] == '1'){
                    count++;
                    q.add(i);
                    q.add(j);
                    grid[i][j] = '0';
                    while(!q.isEmpty()){
                        int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};
                        int rc = q.poll(); int cc = q.poll();
                        for(int[] dir: dirs){
                            int r = dir[0] + rc;
                            int c = dir[1] + cc;
                            if(r >=0 && r < m && c>=0 && c < n && grid[r][c] == '1'){
                                grid[r][c] = '0';
                                q.add(r);
                                q.add(c);
                            }
                        }
                    }
                }
            }
        }
        return count;
    }
}
*/

File: islandcount.py
class Solution:
    
    """
    Description: Given an m x n 2d grid map of '1's (land) and '0's (water), return the number of islands
    
    Time Complexicity: O(m*n)
    Space Complexicity: O(m*n) for DFS and min(m,n) for BFS
    where m and n are dimensions of the given matrix
    
    Approach: 
    Using BFS:
    - take a queue for indices starting when value == "1", change the value to "0" (avoid revisiting)
    - find all nearby "1" in all directions, add indices to queue again and replace by "0" until queue is empty
    - update count after each of the starting queue becomes empty
    
    Using DFS:
    - for first occuring value == "1", add a count, replace "1" by "0"
    - call dfs recursively by replacing all nearby "1" to "0", update count with every new "1" (after replacement)
    
    """
   
    def numIslands(self, grid: List[List[str]]) -> int:
        
        count = 0
        if grid == None or len(grid) == 0: return count
        self.dirs = [(1,0), (0,1), (-1,0), (0,-1)]
        self.m = len(grid); self.n = len(grid[0])        
        
        for row in range(self.m):
            for col in range(self.n):
                if grid[row][col] == "1":
                    count += 1
                    self.dfs(grid, row, col)
        
        return count
    
    def dfs(self, grid, row, col):
        
        # base
        if row < 0 or col < 0 or row == self.m or col == self.n or grid[row][col] != "1": return
        
        # logic
        grid[row][col] = "0"
        for d in self.dirs:
            r = row + d[0]
            c = col + d[1]
            self.dfs(grid, r, c)
            
    def numIslands(self, grid: List[List[str]]) -> int:
        
        count = 0
        if grid == None or len(grid) == 0: return count
        from collections import deque
        
        queue = deque()
        dirs = [(1,0), (0,1), (-1,0), (0,-1)]
        m = len(grid); n = len(grid[0])
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    count += 1; grid[i][j] = "0"
                    queue.append((i, j))
                    while queue:
                        curr = queue.popleft()
                        for d in dirs:
                            row = curr[0] + d[0]
                            col = curr[1] + d[1]
                            if row >= 0 and col >= 0 and row < m and col < n and grid[row][col] == "1":
                                grid[row][col] = "0"
                                queue.append((row,col))
                            
        return count
